group/assignment5/AlphabetFromDictionary.cpp

```diff
@@ -1,120 +1,175 @@
-    vector<bool> isCharAdjacent = this->charPriorityGraph[priorChar];
+	visited[priorChar] = true;
+	vector<bool> isCharAdjacent = this->charPriorityGraph[priorChar];
 
-    for (char c = 0; c < isCharAdjacent.size(); c++) {
-        if (isCharAdjacent[c] && !visited[c]) {
-            topologicalSortUtil(c, visited, stack);
-        }
-    }
+	for (char c = 0; c < isCharAdjacent.size(); c++) {
+		if (isCharAdjacent[c] && !visited[c]) {
+			topologicalSortUtil(c, visited, stack);
+		}
+	}
 
-    stack.push(priorChar);
+	stack.push(priorChar);
 }
 
 void AlphabetFromDictionary::topologicalSort() {
-    stack<char> stack;
-    vector<bool> visited(NUM_OF_NATIVE_CHARS, false);
-
-    for (auto it : this->charPriorityGraph) {
-        if (!visited[it.first]) {
-            this->topologicalSortUtil(it.first, visited, stack);
-        }
-    }
-
-    while (!stack.empty()) {
-        this->orderedAlphabet.push_back(stack.top());
-        stack.pop();
-    }
+	stack<char> stack;
+	vector<bool> visited(NUM_OF_NATIVE_CHARS, false);
+
+	for (auto it : this->charPriorityGraph) {
+		if (!visited[it.first]) {
+			this->topologicalSortUtil(it.first, visited, stack);
+		}
+	}
+
+	while (!stack.empty()) {
+		this->orderedAlphabet.push_back(stack.top());
+		stack.pop();
+	}
 }
 
 void AlphabetFromDictionary::findAlphabet() {
-    if (dictionary.empty()) {
-        return;
-    }
-
-    this->createCharPriorityGraph();
-    if (!this->isGraphCyclic) {
-        // find topological sort of characters graph to find ordered alphabet
-        this->topologicalSort();
-    }
+	if (dictionary.empty()) {
+		return;
+	}
+
+	this->createCharPriorityGraph();
+	if (!this->isGraphCyclic) {
+		// find topological sort of characters graph to find ordered alphabet
+		this->topologicalSort();
+	}
+}
+
+// TODO: part of challenge 1
+void AlphabetFromDictionary::alltopologicalSortUtil(set<vector<char> > &allAlphabets,
+	vector<char> &alphabet,
+	vector<bool> &visited){
+	bool notEndOfAlphabet = false;
+
+	for (char alphabetchar : this->alphabetCharacters)
+	{
+		if (this->inDegree[alphabetchar] == 0 && !visited[alphabetchar])
+		{
+			vector<bool> isCharAdjacent = this->charPriorityGraph[alphabetchar];
+			for (char c = 0; c < isCharAdjacent.size(); c++){
+				if (isCharAdjacent[c])
+					this->inDegree[c]--;
+			}
+
+			alphabet.push_back(alphabetchar);
+			visited[alphabetchar] = true;
+			alltopologicalSortUtil(allAlphabets, alphabet, visited);
+
+			visited[alphabetchar] = false;
+			alphabet.erase(alphabet.end() - 1);
+			for (char c = 0; c < isCharAdjacent.size(); c++){
+				if (isCharAdjacent[c])
+					this->inDegree[c]++;
+			}
+
+			notEndOfAlphabet = true;
+		}
+	}
+	if (!notEndOfAlphabet)
+		allAlphabets.insert(alphabet);
+}
+
+
+// TODO: challenge 1
+set<vector<char> > AlphabetFromDictionary::findAllAlphabets(){
+	set<vector<char> > allAlphabets;
+	if (dictionary.empty()) return allAlphabets;
+
+	vector<bool> visited(NUM_OF_NATIVE_CHARS);

```

Comment:
I still think it's more readable to write "vector visited(NUM_OF_NATIVE_CHARS**, false**);"
Also, I think consistency in the code is important, so either we write "false" in all those declarations or we don't write it in any of them
What do you think?